tools/clang/tools/libclang/dxcrewriteunused.cpp

```diff
@@ -901,19 +999,228 @@ HRESULT DoSimpleReWrite(_In_ DxcLangExtensionsHelper *pHelper,
   if (opts.RWOpt.KeepUserMacro)
-    WriteUserMacroDefines(compiler, o);
+    WriteMacroDefines(astHelper.userMacros, o);
 
   // Flush and return results.
   o.flush();
   w.flush();
 
-  if (compiler.getDiagnosticClient().getNumErrors() > 0)
+  if (astHelper.bHasErrors)
     return E_FAIL;
   return S_OK;
 }
 
+namespace {
+
+void PreprocessResult(CompilerInstance &compiler, LPCSTR pFileName) {
+  // These settings are back-compatible with fxc.
+  clang::PreprocessorOutputOptions &PPOutOpts =
+      compiler.getPreprocessorOutputOpts();
+  PPOutOpts.ShowCPP = 1;           // Print normal preprocessed output.
+  PPOutOpts.ShowComments = 0;      // Show comments.
+  PPOutOpts.ShowLineMarkers = 1;   // Show \#line markers.
+  PPOutOpts.UseLineDirectives = 1; // Use \#line instead of GCC-style \# N.
+  PPOutOpts.ShowMacroComments = 0; // Show comments, even in macros.
+  PPOutOpts.ShowMacros = 0;        // Print macro definitions.
+  PPOutOpts.RewriteIncludes = 0;   // Preprocess include directives only.
+
+  FrontendInputFile file(pFileName, IK_HLSL);
+  clang::PrintPreprocessedAction action;
+  if (action.BeginSourceFile(compiler, file)) {
+    action.Execute();
+    action.EndSourceFile();
+  }
+}
+
+class RewriteVisitor : public RecursiveASTVisitor<RewriteVisitor> {
+public:
+  RewriteVisitor(Rewriter &R, TranslationUnitDecl *tu, RewriteHelper &helper)
+      : TheRewriter(R), SourceMgr(R.getSourceMgr()), tu(tu),
+        helper(helper), bNeedLineInfo(false) {}
+
+  bool VisitFunctionDecl(FunctionDecl *f) {
+    if (helper.unusedFunctions.count(f)) {
+      bNeedLineInfo = true;
+
+      TheRewriter.RemoveText(f->getSourceRange());
+      return true;
+    }
+
+    AddLineInfoIfNeed(f->getLocStart());
+    return true;
+  }
+
+  bool VisitTypeDecl(TypeDecl *t) {
+    if (helper.unusedTypes.count(t)) {
+      bNeedLineInfo = true;
+      TheRewriter.RemoveText(t->getSourceRange());
+      return true;
+    }
+    AddLineInfoIfNeed(t->getLocStart());
+    return true;
+  }
+
+  bool VisitVarDecl(VarDecl *vd) {
+    if (vd->getDeclContext() == tu) {
+      if (helper.unusedGlobals.count(vd)) {
+        bNeedLineInfo = true;
+        TheRewriter.RemoveText(vd->getSourceRange());
+        return true;
+      }
+
+      AddLineInfoIfNeed(vd->getLocStart());
+    }
+    return true;
+  }
+
+private:
+  void AddLineInfoIfNeed(SourceLocation Loc) {
+    if (bNeedLineInfo) {
+      bNeedLineInfo = false;
+      auto lineStr = MakeLineInfo(Loc);
+      TheRewriter.InsertTextBefore(Loc, lineStr);
+    }
+  }
+  std::string MakeLineInfo(SourceLocation Loc) {
+    if (Loc.isInvalid())
+      return "";
+    if (!Loc.isFileID())
+      return "";
+
+    PresumedLoc PLoc = SourceMgr.getPresumedLoc(Loc);
+    const char *Filename = PLoc.getFilename();
+    int Line = PLoc.getLine();
+
+    std::string lineStr;
+    raw_string_ostream o(lineStr);
+    o << "#line" << ' ' << Line << ' ' << '"';
+    o.write_escaped(Filename);
```

Comment:
Can't we skip the filename if the previous `#line` directive referenced the same file?  We should also be able to skip the whole thing if the current line would be the correct line and file, based on the last directive plus number of lines since that directive.  This should be useful for sequences of statements, I imagine.